src/mito2/src/sst/parquet/row_group.rs

```diff
@@ -189,6 +196,60 @@ impl<'a> InMemoryRowGroup<'a> {
         Ok(())
     }
 
+    /// Fetches data from object store.
+    /// If the object store supports blocking, use sequence blocking read.
+    /// Otherwise, use concurrent read.
+    async fn get_byte_ranges(&self, ranges: &[Range<usize>]) -> Result<Vec<Bytes>> {
+        let ranges: Vec<_> = ranges
+            .iter()
+            .map(|range| range.start as u64..range.end as u64)
+            .collect();
+        if self.object_store.info().full_capability().blocking {
+            self.get_ranges_seq(ranges).await
+        } else {
+            self.get_ranges_concurrent(ranges).await
+        }
+    }
+
+    /// Fetches data from object store sequentially
+    async fn get_ranges_seq(&self, ranges: Vec<Range<u64>>) -> Result<Vec<Bytes>> {
+        let block_object_store = self.object_store.blocking();
+        let file_path = self.file_path.to_string();
+
+        let f = move || -> Result<Vec<Bytes>> {
+            ranges
+                .into_iter()
+                .map(|range| {
+                    let data = block_object_store
+                        .read_with(&file_path)
+                        .range(range)
+                        .call()
+                        .map_err(|e| ParquetError::External(Box::new(e)))?;
+                    Ok::<_, ParquetError>(Bytes::from(data))
+                })
+                .collect::<Result<Vec<_>>>()
+        };
+
+        maybe_spawn_blocking(f).await
+    }
+
+    /// Fetches data from object store concurrently.
+    async fn get_ranges_concurrent(&self, ranges: Vec<Range<u64>>) -> Result<Vec<Bytes>> {
```

Comment:
Too many small requests could result in an expensive bill; there is an optimization we can do in the future(maybe not in this PR) for object stores like s3. If the `ranges` are almost continuous, we can merge these ranges into a large chunk and fetch this chunk in the preferred size concurrently.

Reply:
I add a TODO for it